Lib/ufo2ft/instructionCompiler.py

```diff
@@ -0,0 +1,267 @@
+                "Ignoring TrueType instructions for this glyph."
+            )
+            return False
+
+        # Check the glyph hash against the TTGlyph that is being built
+
+        ttwidth = self.otf["hmtx"][glyph.name][0]
+        hash_pen = HashPointPen(ttwidth, self.otf.getGlyphSet())
+        ttglyph.drawPoints(hash_pen, self.otf["glyf"])
+
+        if glyph_hash != hash_pen.hash:
+            logger.error(
+                f"The stored hash for glyph '{glyph.name}' does not match the TrueType "
+                "output glyph. Glyph will have no instructions in the font."
+            )
+            return False
+        return True
+
+    def _check_tt_data_format(self, ttdata, name):
+        """Make sure we understand the format version, currently only version 1
+        is supported."""
+        formatVersion = ttdata.get("formatVersion", None)
+        if not isinstance(formatVersion, str):
+            raise TypeError(
+                f"Illegal type '{type(formatVersion).__name__}' instead of 'str' for "
+                f"formatVersion for instructions in {name}."
+            )
+        if formatVersion != "1":
+            raise NotImplementedError(
+                f"Unknown formatVersion {formatVersion} for instructions in {name}."
+            )
+
+    def _compile_program(self, key, table_tag):
+        """Compile the program for prep or fpgm."""
+        assert table_tag in ("prep", "fpgm")
+        ttdata = self.ufo.lib.get(TRUETYPE_INSTRUCTIONS_KEY, None)
+        if ttdata:
+            self._check_tt_data_format(ttdata, f"lib key '{key}'")
+            asm = ttdata.get(key, None)
+            if asm is None:
+                # The optional key is not there, quit right here
+                return
+            if not asm:
+                # If assembly code is empty, don't bother to add the table
+                logger.debug(
+                    f"Assembly for table '{table_tag}' is empty, "
+                    "table not added to font."
+                )
+                return
+
+            self.otf[table_tag] = table = ttLib.newTable(table_tag)
+            table.program = ttLib.tables.ttProgram.Program()
+            table.program.fromAssembly(asm)
+
+    def compileGlyphInstructions(self, ttGlyph, name):
+        """Compile the glyph instructions from the UFO glyph `name` to bytecode
+        and add it to `ttGlyph`."""
+        if name not in self.ufo:
+            # Skip glyphs that are not in the UFO, e.g. '.notdef'
+            logger.info(
+                f"Skipping compilation of instructions for glyph '{name}' because it "
+                "is not in the input UFO."
+            )
+            return
+
+        glyph = self.ufo[name]
+        ttdata = glyph.lib.get(TRUETYPE_INSTRUCTIONS_KEY, None)
+        if ttdata is not None:
+            self._compile_tt_glyph_program(glyph, ttGlyph, ttdata)
+        if ttGlyph.isComposite():
+            # Remove empty glyph programs from composite glyphs
```

Comment:
I can see what the code is doing, a comment if anything should explain _why_ that's needed

Reply:
if ttGlyph.program is present then WE_HAVE_INSTRUCTIONS is set, but anyway this is already handled in _compile_tt_glyph_program().

Reply:
I added that in response to https://github.com/fonttools/fonttools/pull/1844, but @moyogo is right, it is unnecessary because `_compile_tt_glyph_program()` handles it (which was added later than the explicit removal of the empty program here.